topostats/tracing/disordered_tracing.py

```diff
@@ -0,0 +1,708 @@
+
+        Parameters
+        ----------
+        orig_mask : npt.NDArray
+            Original mask.
+        new_mask : npt.NDArray
+            New mask.
+        holearea_min_max : list
+            List of minimum and maximum hole area (in pixels).
+
+        Returns
+        -------
+        npt.NDArray
+            Smoothed mask with holes restored.
+        """
+        holesize_min_px = holearea_min_max[0] / ((self.pixel_to_nm_scaling / 1e-9) ** 2)
+        holesize_max_px = holearea_min_max[1] / ((self.pixel_to_nm_scaling / 1e-9) ** 2)
+        holes = 1 - orig_mask
+        holes = label(holes)
+        sizes = [holes[holes == i].size for i in range(1, holes.max() + 1)]
+        holes[holes == 1] = 0  # set background to 0
+
+        for i, size in enumerate(sizes):
+            if size < holesize_min_px or size > holesize_max_px:  # small holes may be fake are left out
+                holes[holes == i + 1] = 0
+        holes[holes != 0] = 1
+
+        # compare num holes in each mask
+        holey_smooth = new_mask.copy()
+        holey_smooth[holes == 1] = 0
+
+        return holey_smooth
+
+    @staticmethod
+    def remove_touching_edge(skeleton: npt.NDArray) -> npt.NDArray:
+        """
+        Remove any skeleton points touching the border (to prevent errors later).
+
+        Parameters
+        ----------
+        skeleton : npt.NDArray
+            A binary array where touching clusters of 1's become 0's if touching the edge of the array.
+
+        Returns
+        -------
+        npt.NDArray
+            Skeleton without points touching the border.
+        """
+        for edge in [skeleton[0, :-1], skeleton[:-1, -1], skeleton[-1, 1:], skeleton[1:, 0]]:
+            uniques = np.unique(edge)
+            for i in uniques:
+                skeleton[skeleton == i] = 0
+        return skeleton
+
+    def smooth_mask(
+        self, grain: npt.NDArray, dilation_iterations: int = 2, gaussian_sigma: float | int | None = None
+    ) -> npt.NDArray:
+        """
+        Smooth grains based on the lower number of binary pixels added from dilation or gaussian.
+
+        This method ensures gaussian smoothing isn't too aggressive and covers / creates gaps in the mask.
+
+        Parameters
+        ----------
+        grain : npt.NDArray
+            Numpy array of the grain mask.
+        dilation_iterations : int
+            Number of times to dilate the grain to smooth it. Default is 2.
+        gaussian_sigma : float | None
+            Gaussian sigma value to smooth the grains after an Otsu threshold. If None, defaults to
+            max(grain.shape) / 256.
+
+        Returns
+        -------
+        npt.NDArray
+            Numpy array of smmoothed image.
+        """
+        gaussian_sigma = max(grain.shape) / 256 if gaussian_sigma is None else gaussian_sigma
+        print("-------", dilation_iterations, type(dilation_iterations))
```

Comment:
We should remove all `print()` statements or convert to `LOGGER.info()` with informative messages as to the event being logged.

Reply:
Done